File: src/com/company/stacksandqueues/MinStack.java
package com.company.stacksandqueues;

import com.company.misc.StackNode;

import java.util.Arrays;


public class MinStack {
    int capacity;
    StackNode[] nodes;
    int top;
    int minSoFar = Integer.MAX_VALUE;

    public MinStack(int capacity, StackNode[] nodes) {
        this.capacity = capacity;
        this.nodes = nodes;
    }

    MinStack() {
        top = -1;
        capacity = 30;
        nodes = new StackNode[30];
    }

    boolean isEmpty() {
        return top < 0;
    }

    public void push(int value) {
        if (top >= capacity) {
            System.out.println("Stack Overflow");
        } else {
            int i = ++top;
            nodes[i] = new StackNode();
            if (value < minSoFar) {
                nodes[i].minSoFar = value;
                minSoFar = value;
            } else {
                nodes[i].minSoFar = minSoFar;
            }
            nodes[i].data = value;
        }
    }


    public StackNode pop() {
        if (top == -1) {
            return null;
        } else {
            return nodes[top--];
        }
    }

    @Override
    public String toString() {
        return "Stack{" +
                "capacity=" + capacity +
                ", nodes=" + Arrays.toString(nodes) +
                ", top=" + top +
                ", minSoFar=" + minSoFar +
                '}';
    }
}

File: src/com/company/gfg/stack/MinStack.kt
package com.company.gfg.stack

fun main() {

    //* Your MinStack object will be instantiated and called as such:
    val obj = MinStack()
    obj.push(-2)
    obj.push(0)
    obj.push(-3)
    println(obj.getMin())
    obj.pop()
    println(obj.top())
    println(obj.getMin())


}

class MinStack {

    class StackNode(
            var data: Int,
            var minSoFar: Int
    )

    /** initialize your data structure here. */
    var nodes = arrayListOf<StackNode>()
    var top = -1
    var minSoFar = Int.MAX_VALUE

    fun push(x: Int) {
        minSoFar = Math.min(minSoFar, x)
        nodes.add(++top, StackNode(x, minSoFar))
    }

    fun pop() {
        if (top != -1) {
            nodes.removeAt(top--)
        }
    }

    fun top(): Int {
        return nodes.get(top).data
    }

    fun getMin(): Int {
        return nodes.get(top).minSoFar
    }

}
